### PR TITLE
Add command to support making debian packages from a tar.gz input

### PR DESCRIPTION
**Pull request by kminehart**
The 'deb' subcommand allows us to make a debian package from a tar.gz created by the 'package' command.

To use:

1. Make a tar.gz: `dagger run go run ./cmd package publish --destination=file://dist --verbose --distro=linux/amd64`
2. Use the tar.gz as input to create a deb: `dagger run go run ./cmd deb --destination=file://dist --verbose --package=file://dist/grafana_10.0.0-pre_linux_amd64_*.tar.gz`.

### CODE REVIEW

**Review by ricky-undeadcoders — APPROVED** [review]

**Review by zerok — APPROVED** [review]

File: containers/publish.go [hunk 1]
```diff
@@ -32,34 +31,15 @@ type PublishOpts struct {
...
-			WithExec([]string{"/bin/sh", "-c", "echo $GCP_SERVICE_ACCOUNT_KEY_BASE64 | base64 -d > /key.json"}).
-			File("/key.json")
-
-		auth = NewGCPServiceAccountWithFile(file)
-	} else if key := opts.GCSOpts.ServiceAccountKey; key != "" {
-		log.Println("Handling GCP authentication using Service account key from file...")
-		auth = NewGCPServiceAccount(key)
-	}
-
+	auth := GCSAuth(d, opts.GCSOpts)
```
Comment (guicaulada) [inline containers/publish.go:42]: Glad you moved all the auth stuff to a single function!

File: pipelines/package_names.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+type TarFileOpts struct {
+	Version      string
+	BuildID      string
+	IsEnterprise bool
+	Distro       executil.Distribution
+}
+
+// TarFilename returns a file name that matches this format: {grafana|grafana-enterprise}_{version}_{os}_{arch}_{build_number}.tar.gz
+func TarFilename(opts TarFileOpts) string {
```
Comment (guicaulada) [inline pipelines/package_names.go:18]: Nice move here, I actually have been thinking of how to reduce arguments to functions that take too many args and using an Opts struct type seems like a great idea.

File: slices/contains.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package slices
+
+func Contains[T comparable](value T, arr []T) bool {
```
Comment (guicaulada) [inline slices/contains.go:3]: Nice to see someone using generics in Go.

**Review by guicaulada — APPROVED** [review]
Only some nit picks, small comments and questions.

LGTM!

File: containers/package_input.go [hunk 4]
```diff
@@ -1,5 +1,63 @@
...
+	"dagger.io/dagger"
+)
+
 type PackageInputOpts struct {
-	Package string
+	Packages []string
+	GCSOpts  *GCSOpts
+}
+
+// GetPackage uses the PackageInputOpts to get a Grafana package, either from the local filesystem (if the package is of type 'file://...')
```
[the remark this answers is not in the document]
Comment (kminehart) [reply containers/package_input.go:18]: yeah I think that makes sense and I've already been annoyed by having to specify 'file://' more than once.

File: containers/package_input.go [hunk 5]
```diff
@@ -1,5 +1,63 @@
...
+		}
+
+		files[i] = file
+	}
+
+	return files, nil
+}
+
+func getLocalPackage(ctx context.Context, d *dagger.Client, file string) (*dagger.File, error) {
+	// pending https://github.com/dagger/dagger/issues/4745
```
[the remark this answers is not in the document]
Comment (kminehart) [reply containers/package_input.go:56]: maybe. The code is a little confusing because it's all generated.

File: pipelines/package_names.go (hunk 2 above)
Comment (kminehart) [reply pipelines/package_names.go:18]: One drawback is that if we add fields to the the opts struct the compiler won't tell us in places where we forgot to populate them.

**Pull request merged**